### PR TITLE
Subaru pre-global safety refactor

### PR DESCRIPTION
**Pull request by martinl**
This PR adds safety parameter for models with flipped (scaling factor -1) driver torque signal, for 2017-2018 Forester and 2018-2019 Outback. Also adds safety replay tests for Subaru Forester and Subaru Legacy.

### CODE REVIEW

**Comment by adeebshihadeh** [comment]
Let’s just add the route for the forester since it has the flipped torque signal. We’re going to expand safety replay to test all the supported cars soon.

File: board/safety/safety_subaru.h [hunk 1]
```diff
@@ -35,6 +35,9 @@ AddrCheckStruct subaru_l_rx_checks[] = {
 };
 const int SUBARU_L_RX_CHECK_LEN = sizeof(subaru_l_rx_checks) / sizeof(subaru_l_rx_checks[0]);
 
+const uint16_t SUBARU_PARAM_FLIP_DRIVER_TORQUE = 1;
+bool subaru_flip_driver_torque = false;
```
Comment (adeebshihadeh) [inline board/safety/safety_subaru.h:39]: Let's specify that it's legacy. Maybe we should give legacy it's own file?

File: tests/safety_replay/test_safety_replay.py [hunk 2]
```diff
@@ -23,6 +23,8 @@
   ("5b7c365c50084530_2020-04-15--16-13-24--3--rlog.bz2", Panda.SAFETY_HYUNDAI, 0),  # HYUNDAI.SONATA
   ("610ebb9faaad6b43|2020-06-13--15-28-36.bz2", Panda.SAFETY_HYUNDAI_LEGACY, 0),  # HYUNDAI.IONIQ_EV_LTD
   ("5ab784f361e19b78_2020-06-08--16-30-41.bz2", Panda.SAFETY_SUBARU_LEGACY, 0),  # SUBARU.OUTBACK
+  ("95441c38ae8c130e_2020-06-08--12-10-17.bz2", Panda.SAFETY_SUBARU_LEGACY, 1),  # SUBARU.FORESTER
```
Comment (adeebshihadeh) [inline tests/safety_replay/test_safety_replay.py:26]: `SUBARU.FORESTER` is the global model

File: board/safety/safety_subaru.h (hunk 1 above)
Comment (martinl) [reply board/safety/safety_subaru.h]: Renamed the variable and moved legacy to separate file. 2020 global hybrid will also have separate safety mode so it made sense to separate legacy now

File: tests/safety_replay/test_safety_replay.py (hunk 2 above)
Comment (martinl) [reply tests/safety_replay/test_safety_replay.py]: changed the comments to match preglobal PR labels

**Comment by martinl** [comment]
looks like safety replay fails due to missing 5ab784f361e19b78_2020-06-08--16-30-41.bz2

File: board/safety/safety_subaru_legacy.h [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    // sample wheel speed, averaging opposite corners
+    if (addr == 0xD4) {
+      int subaru_speed = (GET_BYTES_04(to_push) >> 16) & 0xFFFF;  // FR
+      subaru_speed += GET_BYTES_48(to_push) & 0xFFFF;  // RL
+      subaru_speed /= 2;
+      vehicle_moving = subaru_speed > SUBARU_L_STANDSTILL_THRSLD;
+    }
+
+    if (addr == 0xD1) {
+      brake_pressed = ((GET_BYTES_04(to_push) >> 16) & 0xFF) > 0;
```
Comment (adeebshihadeh) [inline board/safety/safety_subaru_legacy.h:67]: use GET_BYTE instead of shifting and anding

File: board/safety/safety_subaru_legacy.h [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+      torque_driver_new = to_signed(torque_driver_new, 11);
+      if (subaru_l_flip_driver_torque) {
+        torque_driver_new = -1 * torque_driver_new;
+      }
+      update_sample(&torque_driver, torque_driver_new);
+    }
+
+    // enter controls on rising edge of ACC, exit controls on ACC off
+    if (addr == 0x144) {
+      int cruise_engaged = ((GET_BYTES_48(to_push) >> 17) & 1);
```
Comment (adeebshihadeh) [inline board/safety/safety_subaru_legacy.h:48]: you're only using one byte here, use GET_BYTE instead

File: board/safety/safety_subaru_legacy.h (hunk 3 above)
Comment (martinl) [reply board/safety/safety_subaru_legacy.h]: done

File: board/safety/safety_subaru_legacy.h (hunk 4 above)
Comment (martinl) [reply board/safety/safety_subaru_legacy.h]: also done

**Comment by martinl** [comment]
I rebased to master and changed driver torque 75 > 60 to match Global

**Comment by martinl** [comment]
safety replay is currently failing because preglobal Forester route (95441c38ae8c130e_2020-06-08--12-10-17.bz2) is missing from server

**Comment by martinl** [comment]
I added a brake pressed threshold to carstate and panda safety which fixes disengage while ACC braking issue on some preglobal cars https://github.com/martinl/openpilot/issues/27

**Comment by martinl** [comment]
I will close this and open new PR with fresh rebase

**Pull request closed without merging**